kv_store/include/server_utility.h

```diff
@@ -57,20 +56,39 @@ bool responsible(string key, int rep, consistent_hash_map<N,H>& hash_ring, strin
   }
 
 
-  // at the end of the previous loop, pos has the last node responsible for the
-  // key; we only remove that key if we have more nodes than there should be replicas
-  if (resp && (hash_ring.size() > rep)) {
-    remove = true;
-    sender_node = pos->second;
-  } else if (resp && (hash_ring.size() <= rep)) {
-    remove = false;
-    if (++target_pos == hash_ring.end()) {
-      target_pos = hash_ring.begin();
-    }
+  // at the end of the previous loop, pos has the last node responsible for the key
+  if (resp) {
+    return pair<bool, N*>(resp, &(pos->second));
+  } else {
+    return pair<bool, N*>(resp, nullptr);
+  }
+}
 
-    sender_node = target_pos->second;
+template<typename T>
+communication::Key_Response get_key_address(string target_node_address, string target_tier, unordered_set<string> keys, SocketCache& key_address_requesters, unordered_map<string, T> placement) {
+  // form key address request
+  communication::Key_Request req;
+  req.set_sender("server");
+  if (target_tier == "M") {
```

Comment:
This code doesn't make sense to me... why don't we just do `req.set_target_tier(target_tier)`?

Reply:
Oh that's because this function is also called during intra-tier gossip, in which case we don't have a target tier. Currently in these calls, the "target_tier" is left as an empty string.